### PR TITLE
Various docs improvements from JOSS review

### LINKED ISSUE #128: Add "Contributing" section to the Readme.md
Following one of the points in the checklist for JOSS
https://github.com/openjournals/joss-reviews/issues/5669#issuecomment-1667955949

> Community guidelines: Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

I suggest you to add a section to the Readme.md file similar to what Oceananigans.jl have.
https://github.com/CliMA/Oceananigans.jl#contributing

### LINKED ISSUE #129: State of the field
Following one of the points on the checklist
https://github.com/openjournals/joss-reviews/issues/5669#issuecomment-1667955949

> State of the field: Do the authors describe how this software compares to other commonly-used packages?

I think that it should be added to the manuscript a description on how biogeochemical models are currently performed using Oceananigans.jl and how this package is improving the current state of the field.

Even myself have already used Oceananigans.jl coupled with a simple NP model, which led to a published paper,
https://www.frontiersin.org/articles/10.3389/fmars.2022.825027/full
https://github.com/iuryt/NorthAtlanticBloom
and I developed a very simple package that can be used for estimating the mixed-layer depth and light-limiting growth.
https://github.com/iuryt/Bioceananigans.jl

While those were a contribution to the field, I clearly see a huge improvement that your package is making to the state of the field, which I don't think you are making it clear in the manuscript.

## Advantages:

This is a long list, but in summary, this package is clearly well integrated with Oceananigans.jl and have many options for different models both using tracers and particles. It has a documentation and description of the different variables.

## Disadvantages (if any):

This is kind of hard to find a disadvantage for this package in relation to the current state of the field, but I will try to list something here.
The first one is that there is no way to change the equations or define my own set of equations for my biogeochemical model. This is important for theoretical process modelling approches.
I also could not find any way to define the mixed layer from the fields (there is a discussion to [add that to Oceanostics.jl](https://github.com/tomchor/Oceanostics.jl/issues/79)) and there are actually different ways to deal with the effect of the mixed layer on the biogeochemical modelling in the literature. The main point is that there are mismatches between the time scales of the mixing and cell division. Although light-averaging within the mixed layer has been previously employed and compared to observations (Mahadevan et al., 2012), other studies approach this issue differently (e.g., Sarmiento et al., 1993). Fasham et al., (1990), for example, average the light-limiting growth in the mixed layer. I could not see an implementation for that in this package.

While I don't think this is something you need to implement for the publication, it is important to, at least, mention that in the manuscript. If needed, I can add another issue for this improvement for tracking purposes.

## References

Mahadevan A., D’asaro E., Lee C., Perry M. J. (2012). Eddy-driven stratification initiates north atlantic spring phytoplankton blooms. Science 337, 54–58. doi: 10.1126/science.1218740

Sarmiento J.L., Slater R.D., Fasham M.J.R., Ducklow H.W., Toggweiler J.R., Evans G.T. (1993) A seasonal three‐dimensional ecosystem model of nitrogen cycling in the North Atlantic euphotic zone. Global biogeochemical cycles, 7 (2), 417–450. doi: 10.1029/93GB00375

Fasham M. J., Ducklow H. W., McKelvie S. M. (1990). A nitrogen-based model of plankton dynamics in the oceanic mixed layer. J. Mar. Res. 48, 591–639. doi: 10.1357/002224090784984678

### PR DESCRIPTION
**Pull request by jagoosw**
This PR addresses several docs problems from the [JOSS review](https://github.com/openjournals/joss-reviews/issues/5669)

This includes

- [x] Adds model equations for all model components
- [x] Adds parameter variable name lists to each model component
- [x] Speeds up quickstart example by reducing resolution (and makes the example more realistic)
- [x] Adds flow diagrams explaining the bgc model formulations
- [x] Adds contributors guide (closes #128)
- [ ] Adds docs section explaining how to implement new models (closes #129)

### CODE REVIEW

**Comment by jagoosw** [comment]
I can't work out how to write a piecewise equation nicely in Documenters latex. In normal latex you can write `\left\{ \begin{array}...\end{array}\right` but this doesn't work in Documenter

**Comment by navidcy** [comment]
> I can't work out how to write a piecewise equation nicely in Documenters latex. In normal latex you can write `\left\{ \begin{array}...\end{array}\right` but this doesn't work in Documenter

Try escaping \ chars, eg

\\begin{array}…

**Comment by jagoosw** [comment]
> > I can't work out how to write a piecewise equation nicely in Documenters latex. In normal latex you can write `\left\{ \begin{array}...\end{array}\right` but this doesn't work in Documenter
> 
> Try escaping \ chars, eg
> 
> ```
> \\begin{array}…
> ```

Think I've got this working now

**Review by navidcy — CHANGES_REQUESTED** [review]
There is something wrong w rendering at

https://oceanbiome.github.io/OceanBioME.jl/previews/PR126/model_components/air-sea-gas/

**Comment by jagoosw** [comment]
> There is something wrong w rendering at
> 
> https://oceanbiome.github.io/OceanBioME.jl/previews/PR126/model_components/air-sea-gas/

Good spot, thank you

**Comment by navidcy** [comment]
Can we try size = (160, 32) for the first example? Movie looks bit pixelated now.

File: README.md
Comment (navidcy) [inline README.md:1]: We also need to update the video link to the one made with the updated resolution.

**Comment by jagoosw** [comment]
> Can we try size = (160, 32) for the first example? Movie looks bit pixelated now.

Sure, seems reasonable

File: README.md
Comment (jagoosw) [reply README.md:1]: We could probably update it to show the one generated in the docs now instead of being a manual upload

**Comment by jagoosw** [comment]
Todo:
- [ ] Add lists of parameter variable names to equations listings
- [ ] Add flow diagrams

File: README.md
Comment (jagoosw) [reply README.md:1]: I had a think about this and don't think there's a way for us to get it from the correct branch/preview docs. Do you think we should change it to always get the video from the stable docs? This would only be an issue if we change the video and want to preview it

Comment (navidcy) [reply README.md:1]: Yes! From the stable is better. Or the dev version of docs

**Review by navidcy — APPROVED** [review]
lgtm. I admit I didn’t check the equations (eg for sign errors and what not). 

I saw a todo list. If you want me to have a look at some point ping me.

**Comment by jagoosw** [comment]
I've added tables to each page with equations listing the corresponding variable name for each parameter and added flow diagrams as per the review suggestions (will see how they look).

I also listed the units which we didn't have anywhere before.

**Comment by jagoosw** [comment]
@navidcy I think I've finished making the changes here now. The README image doesn't appear to be loading from the preview which is a bit odd. I set it to the preview branch to test (intending to change to main once it worked) but it's not working.

File: docs/src/index.md [hunk 1]
```diff
@@ -90,14 +90,14 @@ hm2 = heatmap!(ax2, xP, zP, Pₙ, colorrange = P_lims, colormap = Reverse(:bamak
...
-record(fig, "buoyancy_front.mp4", 1:length(times)) do i
+record(fig, "buoyancy_front.gif", 1:length(times)) do i
     n[] = i
 end
 
 nothing #hide
 ```
 
-![buoyancy_front](buoyancy_front.mp4)
+![buoyancy_front](buoyancy_front.gif)
```
Comment (navidcy) [inline docs/src/index.md:100]: Just note that file size is bigger and this will have an imprint of the github repo size

**Comment by navidcy** [comment]
> @navidcy I think I've finished making the changes here now. The README image doesn't appear to be loading from the preview which is a bit odd. I set it to the preview branch to test (intending to change to main once it worked) but it's not working.

Revert back to previous way of showing the video and leave this for a future PR? Upload the latest version of the gif in an issue and update the link?

File: docs/src/index.md (hunk 1 above)
Comment (jagoosw) [reply docs/src/index.md:100]: I'll change this back, I don't think its possible to have videos in the readme so thought this might work to fix it, but doesn't seem to.

File: README.md [hunk 2]
```diff
@@ -28,13 +28,15 @@ julia> Pkg.add("OceanBioME")
 ```
 
 ## Running your first model
-As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front:
+As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front. This also requires Oceananigans, so we will add that first:
```
Comment (navidcy) [inline README.md:31]:
```suggestion
As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front. This example requires Oceananigans, so we install that first:
```

**Comment by jagoosw** [comment]
I've changed to a `SeawaterBuoyancy` with constant salinity to reduce the confusion of also having b. It takes longer to run now though as I had to make the timestep smaller, we could reduce the buoyancy gradient which would probably make it so the timestep can be bigger?

File: docs/src/index.md [hunk 3]
```diff
@@ -25,54 +25,62 @@ julia> Pkg.add("OceanBioME")
 
 ## Running your first model
 
-As a simple example let's run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front:
+As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front. This also requires Oceananigans, so we will add that first:
```
Comment (navidcy) [inline docs/src/index.md:28]:
```suggestion
As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front. This example requires Oceananigans, so we install that first:
```

**Comment by jagoosw** [comment]
Just tried changing the temp gradient and it looks fine still
buoyancy_front
So will change this and reduce the time step again

File: docs/src/index.md [hunk 4]
```diff
@@ -25,54 +25,62 @@ julia> Pkg.add("OceanBioME")
...
 
-```@example quickstart
-using OceanBioME
-using Oceananigans, Oceananigans.Units
+```julia
```
Comment (navidcy) [inline docs/src/index.md:30]: Why did you change @example to simple code? That won’t work. Documenter wont run the code but rather will just print it.

Comment (jagoosw) [reply docs/src/index.md:30]: Ah I didn't mean todo that

Comment (navidcy) [reply docs/src/index.md:30]: (There is a chance I missed something here?)

Comment (jagoosw) [reply docs/src/index.md:30]: Think its because I copied from readme

**Comment by navidcy** [comment]
> Just tried changing the temp gradient and it looks fine still
> buoyancy_front
> So will change this and reduce the time step again

Let’s not use a divergent colormap for T

File: docs/src/index.md (hunk 4 above)
Comment (navidcy) [reply docs/src/index.md]: Cool. Okie!

**Comment by jagoosw** [comment]
buoyancy_front

**Comment by navidcy** [comment]
> Happy with this one?
> buoyancy_front

Magnificent

**Comment by jagoosw** [comment]
buoyancy_front

Maybe it's less pretty but a much larger domain like this might be a more realistic first example? And it runs in about the same time.

Obviously not really a realistic problem but the phytoplankton growth has a chance to develop and be effected by the mixing here?

**Comment by jagoosw** [comment]
buoyancy_front
Improved (?) version with interpolation

**Comment by jagoosw** [comment]
Or maybe something like this where the cold water is nutrient-rich, kind of in the vein of this paper https://www.math.hkust.edu.hk/~magan/AMCE6082010/paperpool2010/Franks_1992.pdf

buoyancy_front

**Comment by iuryt** [comment]
> Or maybe something like this where the cold water is nutrient-rich, kind of in the vein of this paper https://www.math.hkust.edu.hk/~magan/AMCE6082010/paperpool2010/Franks_1992.pdf buoyancy_front

It is very cool to see the effect of the downwelling at the front bringing phytoplankton to greater depths.
I like this version, also because you also show the evolution of the nutrients.

**Comment by jagoosw** [comment]
> A thought: This PR has become large. Why not open new PR for contribution guide etc?

I've already done the contribution guide but forgot to push, so think I will add that and then do the other bit in a new PR?

**Comment by jagoosw** [comment]
Happy for this PR to be merged now?

**Comment by navidcy** [comment]
Merge!
